OrbitVulkanLayer/VulkanLayerController.h

```diff
@@ -0,0 +1,460 @@
+      bool is_unique = true;
+      for (const auto& other_extension : extensions) {
+        if (strcmp(extension.extensionName, other_extension.extensionName) == 0) {
+          is_unique = false;
+          break;
+        }
+      }
+      if (is_unique) {
+        extensions.push_back(extension);
+      }
+    }
+
+    // As above, if properties is nullptr, only the number if extensions is queried.
+    if (properties == nullptr) {
+      *property_count = extensions.size();
+      return VK_SUCCESS;
+    }
+
+    uint32_t num_extensions_to_copy = extensions.size();
+    // In the case that less extensions are queried then the layer uses, we copy on this number and
+    // return VK_INCOMPLETE, according to the specification.
+    if (*property_count < num_extensions_to_copy) {
+      num_extensions_to_copy = *property_count;
+    }
+    memcpy(properties, extensions.data(), num_extensions_to_copy * sizeof(VkExtensionProperties));
+    *property_count = num_extensions_to_copy;
+
+    if (num_extensions_to_copy < extensions.size()) {
+      return VK_INCOMPLETE;
+    }
+    return VK_SUCCESS;
+  }
+
+  [[nodiscard]] const DispatchTable* dispatch_table() const { return &dispatch_table_; }
+
+  [[nodiscard]] const SubmissionTracker* submission_tracker() const { return &submission_tracker_; }
+
+  [[nodiscard]] const DeviceManager* device_manager() const { return &device_manager_; }
+
+  [[nodiscard]] const TimerQueryPool* timer_query_pool() const { return &timer_query_pool_; }
+
+  [[nodiscard]] const QueueManager* queue_manager() const { return &queue_manager_; }
+
+ private:
+  void InitVulkanLayerProducerIfNecessary() {
+    absl::MutexLock lock{&vulkan_layer_producer_mutex_};
+    if (vulkan_layer_producer_ == nullptr) {
+      vulkan_layer_producer_ = std::make_unique<VulkanLayerProducerImpl>();
+      // TODO: As soon as we do have producer side channel for the port, use it.
+      LOG("Bringing up VulkanLayerProducer");
+      vulkan_layer_producer_->BringUp(nullptr);
+      submission_tracker_.SetVulkanLayerProducer(vulkan_layer_producer_.get());
+    }
+  }
+
+  void CloseVulkanLayerProducerIfNecessary() {
+    absl::MutexLock lock{&vulkan_layer_producer_mutex_};
+    if (vulkan_layer_producer_ != nullptr) {
+      // TODO: Only do this when DestroyInstance has been called a number of times
+      //  equal to the number of times CreateInstance was called.
+      LOG("Taking down VulkanLayerProducer");
+      vulkan_layer_producer_->TakeDown();
+      vulkan_layer_producer_.reset();
+      submission_tracker_.SetVulkanLayerProducer(nullptr);
+    }
+  }
+
+  std::unique_ptr<VulkanLayerProducer> vulkan_layer_producer_ = nullptr;
+  absl::Mutex vulkan_layer_producer_mutex_;
+
+  DispatchTable dispatch_table_;
+  DeviceManager device_manager_;
+  TimerQueryPool timer_query_pool_;
+  SubmissionTracker submission_tracker_;
+  QueueManager queue_manager_;
+
+  static constexpr uint32_t kNumTimerQuerySlots = 65536;
```

Comment:
Is this a constant from the specification?

Reply:
Can you add a comment how this is chosen and why this value was chosen? It's fine to say that the choice is "arbitrary but large enough" or something like that, but someone not familiar with the code should have an idea where this value came frame (and knowing that it just needs to be "large enough" is valuable information).

Reply:
Done.